### PR TITLE
Router Refactor 1: Install `react-router-dom` v6 and refactor main app router

### PR DESCRIPTION
**Pull request by kevinhaube**
> This is one part of a larger refactor. This PR merges into a shadow-master branch: `experience/shadow/router-refactor`

This PR makes a few changes in the adoption of react-router v6:

- AuthRoute pattern is no longer supported; now use `AuthElement` to authenticate a user before loading a page.
- Deprecation of `useHistory` and `RouteComponentProps`; now use `useNavigate` and `useParams`. (**NOTE**: We can now more easily throw Errors if any of the params come up undefined -- great perk of the pattern)

Additionally, some refactors lead to added capabilities:

- `AuthElement` supports one or many `MemberType` arguments for auth
- `AuthElement` supports a feature flag
- Throws on components expecting url params but receiving undefined (negates the need for undefined checks in code, AND plays into our error boundary tech-debt)

Test Steps:
1. Pull this branch and run `yarn` to install new packages
2. Start the front-end
3. The home page should load. That is the extend of the work in _this_ PR.

## Changes
- Pages requiring auth now all export `[name]WithAuth` for use in routing
- `AppRouter` extracted from `App` and refactored to meet v6 requirements

### Testing
- [X] Tested locally?
- [ ] Ran `./prime test` or `./gradlew testSmoke` against local Docker ReportStream container?
- [X] (For Changes to /frontend-react/...) Ran `npm run lint:write`? 
- [ ] Added tests?

### Process
- [ ] Are there licensing issues with any new dependencies introduced?
- [ ] Includes a summary of what a code reviewer should test/verify?
- [ ] Updated the release notes?
- [ ] Database changes are submitted as a separate PR?
- [ ] DevOps team has been notified if PR requires ops support?

## Linked Issues
- Fixes #issue

## To Be Done
*Create GitHub issues to track the work remaining, if any*
- #issue

## Specific Security-related subjects a reviewer should pay specific attention to
- Does this PR introduce new endpoints?
    - new endpoint A
    - new endpoint B
- Does this PR include changes in authentication and/or authorization of existing endpoints?
- Does this change introduce new dependencies that need vetting?
- Does this change require changes to our infrastructure?
- Does logging contain sensitive data?
- Does this PR include or remove any sensitive information itself?

If you answered '_yes_' to any of the questions above, conduct a detailed Review that addresses at least:

- What are the potential security threats and mitigations? Please list the _STRIDE_ threats and how they are mitigated
    - **S**poofing (faking authenticity)
        - Threat _T_, which could be achieved by _A_, is mitigated by _M_
    - **T**ampering (influence or sabotage the integrity of information, data, or system)
    - **R**epudiation (the ability to dispute the origin or originator of an action)
    - **I**nformation disclosure (data made available to entities who should not have it)
    - **D**enial of service (make a resource unavailable)
    - **E**levation of Privilege (reduce restrictions that apply or gain privileges one should not have)
- Have you ensured logging does not contain sensitive data?
- Have you received any additional approvals needed for this change?

### CODE REVIEW

File: frontend-react/src/components/Admin/EditReceiverSettings.tsx [hunk 1]
```diff
@@ -343,36 +345,47 @@ const EditReceiverSettingsForm: React.FC<EditReceiverSettingsFormProps> = ({
...
-export function EditReceiverSettings({ match }: RouteComponentProps<Props>) {
-    const orgname = match?.params?.orgname || "";
-    const receivername = match?.params?.receivername || "";
-    const action = match?.params?.action || "";
+export function EditReceiverSettings() {
+    const { orgName, receiverName, action } =
+        useParams<EditReceiverSettingsProps>();
+    /* useParams now returns possible undefined. This will error up to a boundary
+     * if the url param is undefined */
+    if (
```
Comment (kevinhaube) [inline frontend-react/src/components/Admin/EditReceiverSettings.tsx:359]: Might behoove us later to refactor this into some kind of `useValidParams` hook that throws whenever expected `!=` received params (i.e. middleware undefined check)

File: frontend-react/src/components/Admin/EditReceiverSettings.tsx [hunk 2]
```diff
@@ -343,36 +345,47 @@ const EditReceiverSettingsForm: React.FC<EditReceiverSettingsFormProps> = ({
...
                 <Title
-                    preTitle={`Org name: ${
-                        match?.params?.orgname || "missing param 'orgname'"
-                    }`}
-                    title={`Receiver name: ${
-                        match?.params?.receivername ||
-                        "missing param 'receivername'"
-                    }`}
+                    preTitle={`Org name: ${orgName}`}
+                    title={`Receiver name: ${receiverName}`}
```
Comment (kevinhaube) [inline frontend-react/src/components/Admin/EditReceiverSettings.tsx:371]: Exemplary of cleaner code via the above pattern

File: frontend-react/src/components/Admin/EditReceiverSettings.tsx [hunk 3]
```diff
@@ -343,36 +345,47 @@ const EditReceiverSettingsForm: React.FC<EditReceiverSettingsFormProps> = ({
...
-                receivername={receivername}
+                orgname={orgName}
+                receivername={receiverName}
                 action={action}
             />
         </AdminFormWrapper>
     );
 }
+
+export function EditReceiverSettingsWithAuth() {
```
Comment (kevinhaube) [inline frontend-react/src/components/Admin/EditReceiverSettings.tsx:384]: Exemplary of higher-order approach to auth component

**Review by kevinhaube — COMMENTED** [review]
Engineer's notes 📓

File: frontend-react/src/components/Content/MarkdownDirectory.tsx [hunk 4]
```diff
@@ -69,12 +69,12 @@ export const GeneratedRoute = ({ dir }: { dir: ContentDirectory }) => {
             <Route
                 key={`${dir.slug}-route`}
                 path={`${dir.slug}`}
-                render={() => (
+                element={
```
Comment (TomNUSDS) [inline frontend-react/src/components/Content/MarkdownDirectory.tsx:72]: Interesting

File: frontend-react/src/pages/admin/AdminLastMileFailures.tsx [hunk 5]
```diff
@@ -32,3 +34,12 @@ export function AdminLastMileFailures() {
         </NetworkErrorBoundary>
     );
 }
+
+export function AdminLMFWithAuth() {
+    return (
+        <AuthElement
+            element={AdminLastMileFailures}
+            requiredUserType={MemberType.PRIME_ADMIN}
```
Comment (TomNUSDS) [inline frontend-react/src/pages/admin/AdminLastMileFailures.tsx:42]: `requiredFeatureFlag` doesn't seem to be required?

File: frontend-react/src/pages/Validate.tsx [hunk 6]
```diff
@@ -21,3 +27,11 @@ const Validate = () => {
 };
 
 export default Validate;
+
+export const ValidateWithAuth = () => (
+    <AuthElement
+        element={Validate}
+        requiredUserType={MemberType.PRIME_ADMIN}
+        requiredFeatureFlag={FeatureFlagName.VALIDATION_SERVICE}
```
Comment (TomNUSDS) [inline frontend-react/src/pages/Validate.tsx:35]: Can there be multiple flags?

**Review by TomNUSDS — COMMENTED** [review]
Is this not going into master?

> [kevinhaube] wants to merge 4 commits into [experience/shadow/router-refactor](https://github.com/CDCgov/prime-reportstream/tree/experience/shadow/router-refactor) from [experience/router-v5-to-v6](https://github.com/CDCgov/prime-reportstream/tree/experience/router-v5-to-v6)

Also a few other comments, I'm going to load up the branch and play with it.

File: frontend-react/src/pages/Validate.tsx (hunk 6 above)
Comment (kevinhaube) [reply frontend-react/src/pages/Validate.tsx:35]: I considered it, but really we associate one flag to one route (like `/validate`, and `/value-sets`)

File: frontend-react/src/pages/admin/AdminLastMileFailures.tsx (hunk 5 above)
Comment (kevinhaube) [reply frontend-react/src/pages/admin/AdminLastMileFailures.tsx:42]: Yeah, neither of the auth params are. I figured we can't be sure that it'll _always_ include a flag or _always_ include a user, so leaving all auth checks as optional means you can choose one, the other, or both to validate before rendering a page.

**Pull request merged**